README.md
# TicTacToe
Play an AI in Tic Tac Toe!

This includes two versions of AI. The first is a minimax algorithm that serves as an unbeatable AI.

The second is my take (no outside algorithms) on a very primitive neural network that works by weighting moves as positive or negative based on results and enters it into a file (history.txt). When searching for a move, the program interprets that file to find what move to do based on past moves - if it hasn't encountered a particular game state, it finds the closest state and uses that as a starting point for the current state.


History.txt
000000000 0(0)1(0)2(0)3(1)4(0)5(0)6(0)7(0)8(0)010000200)


Node.java

/**
 * A java representation of a tree, or more accurately, the nodes on the tree
 * 
 * @author Alex Van Keulen
 * @date Dec 2016
 */
import java.util.List;
import java.util.ArrayList;

public class Node<Integer> {
    int score;
    int[][] data;
    ArrayList<Node<Integer>> children;

    public Node() { 
        children = new ArrayList<Node<Integer>>();
        score = 0;
        data = new int[3][3];
    }

    /**
     * Adds a node to the root node
     *
     * @param data The game board
     */
    public void addNode(int[][] data)
    {
        Node<Integer> node = new Node<Integer>();
        node.score = 0;
        node.data = new int[3][3];
        for(int i = 0; i < data.length; i++)
            node.data[i] = data[i].clone();
        children.add(node);
    }
}



Neural.java

/**
 * Write a description of class Neural here.
 * 
 * @author (your name) 
 * @version (a version number or a date)
 */
import java.util.ArrayList;

public class Neural
{
    FileReadWrite frw;
    int board[][];
    ArrayList<Integer> choices; //parallel
    ArrayList<String> lines;    //arrays

    public Neural()
    {
        frw = new FileReadWrite("History.txt", true);
        board = new int[3][3];
        choices = new ArrayList<Integer>();
        lines = new ArrayList<String>();
    }

    public int getMove(int board[][])
    {
        for(int i = 0; i < 3; i++)
            for(int j = 0; j < 3; j++)
                this.board[i][j] = board[i][j];
        String gameBoard = toStr(board);
        String search = search(gameBoard);
        int choice = 0;
        if(!search.equals(""))
        {
            choice = getChoice(search, board);
            lines.add(search);
            choices.add(choice);
            return choice;
        } else {
            search = closest(gameBoard);
            addLine(gameBoard, search.split(" ")[1]);
            choice = getChoice(search, board);
            lines.add(search);
            choices.add(choice);
            return choice;
        }
    }

    public void addLine(String gameBoard, String regex)
    {
        frw.write(gameBoard + " " + regex);
    }

    public String search(String search)
    {
        return frw.findLine(search);
    }

    public String closest(String line)
    {
        return frw.findClosest(line);
    }

    public int getChoice(String line, int board[][])
    {
        int choice = 0;
        ArrayList<Integer> choices = new ArrayList<Integer>();
        String weights = line.split(" ")[1];
        String nodes[] = weights.split("[)]");
        int index = 0;
        while(board[index / 3][index % 3] != 0)
            index++;
        int largest = Integer.parseInt(nodes[index].split("[(]")[1]);
        choices.add(index);
        for(int i = index; i < 9; i++)
        {
            int weight = Integer.parseInt(nodes[i].split("[(]")[1]);
            if(weight == largest)
            {
                if(board[i / 3][i % 3] == 0)
                    choices.add(i);
            } else if(weight > largest) {
                if(board[i / 3][i % 3] == 0)
                {
                    choices.clear();
                    choices.add(i);
                    largest = weight;
                }
            }
        }
        choice = choices.get((int) Math.floor(Math.random() * choices.size()));
        return choice;
    }

    public void eval(boolean tie, int modifier)
    {
        for(int i = 0; i < lines.size(); i++)
        {
            try{
                frw.replace(lines.get(i), choices.get(i), i*modifier, tie);
            } catch (IndexOutOfBoundsException e) {
                System.out.println("Error");
            }
        }
        choices.clear();
        lines.clear();
    }

    public String toStr(int board[][])
    {
        String ret = "";
        for(int i = 0; i < 3; i++)
            for(int j = 0; j < 3; j++)
            {
                if(board[i][j] == 10)
                    ret += 2;
                else ret += board[i][j];
            }

        return ret;
    }

}


TicTacToe.java

/**
 * O
 * 
 * @author Alex Van Keulen
 * @date Dec 2016
 */
import java.util.ArrayList;

public class TicTacToe
{
    public TicTacToe(){}

    /**
     * Makes a move for the AI by populating a decision tree, and recursively checking the nodes for either A) Max if it is the computers turn, or B) Min if it is the users turn
     *
     * @param root Initially the original board
     * @param player Whoevers turn it is
     * @return Returns either the minimum or maximum nodes of a level, depending on whose turn it is
     */
    public int move(Node<Integer> root, int player)
    {
        player %= 2;
        populate(player * 9 + 1, root); //sends player param as either 1 or 10
        if(root.children.size() == 0) //board is full
            return 0;
        int score = 0;
        for(int i = 0; i < root.children.size(); i++)
        {
            score = check(root.children.get(i).data);
            root.children.get(i).score = score;
            if(score == 0)
            {
                root.children.get(i).score = move(root.children.get(i), player + 1);
            }
        }
        if(player == 0)
            return max(root).score;
        else
            return min(root).score;
    }

    /**
     * This returns the node that has the maximum value in a level
     *
     * @param root The parent of the level
     * @return The node with the maximum value
     */
    public Node<Integer> max(Node<Integer> root)
    {
        int max = -10;
        int j = 0;
        for(int i = 0; i < root.children.size(); i++)
        {
            if(root.children.get(i).score > max)
            {
                max = root.children.get(i).score;
                j = i;
            }
        }
        return root.children.get(j);
    }

    /**
     * This returns the node that has the minimum value in a level
     *
     * @param root The parent of the level
     * @return The node with the minimum value
     */
    public Node<Integer> min(Node<Integer> root)
    {
        int min = 10;
        int j = 0;
        for(int i = 0; i < root.children.size(); i++)
        {
            if(root.children.get(i).score < min)
            {
                min = root.children.get(i).score;
                j = i;
            }
        }
        return root.children.get(j);
    }

    /**
     * Gets the board space of the node that has the maximum value. If two or more nodes have the same max value, it randomly chooses between them
     *
     * @param root The parent of the level
     * @return The board space corresponding to the best move
     */
    public int endGame(Node<Integer> root)
    {
        int max = -10;
        ArrayList<Integer> j = new ArrayList<Integer>();
        for(int i = 0; i < root.children.size(); i++)
        {
            if(root.children.get(i).score >= max)
            {
                if(root.children.get(i).score > max)
                    j.clear();
                max = root.children.get(i).score;
                j.add(i);
            }
        }
        int i = (int) Math.floor(Math.random() * j.size());
        i = j.get(i);
        return getSpace(root.data, root.children.get(i).data);
    }

    /**
     *  Get's the space the node is pointing to by comparing the parent and child node, and seeing which index is different
     *
     * @param arr1 The parent nodes array
     * @param arr2 The child nodes array
     * @return The space the node is referring to
     */
    public int getSpace(int[][] arr1, int[][] arr2)
    {
        for(int i = 0; i < arr1.length; i++)
        {
            for(int j = 0; j < arr2.length; j++)
            {
                if(arr1[i][j] != arr2[i][j])
                    return (i*3 + j);
            }
        }
        return 0;
    }

    /**
     * Populates the root given with all possible moves
     *
     * @param player The player (1 or 10) whose turn it is
     * @param root The root to populate
     */
    public void populate(int player, Node<Integer> root)
    {
        int[][] board = root.data;
        int[][] tempBoard = new int[3][3];
        for(int i = 0; i < board.length; i++)
            tempBoard[i] = board[i].clone();
        for(int i = 0; i < board.length; i++)
        {
            for(int j = 0; j < board.length; j++)
            {
                if(board[i][j] == 0) //empty
                {
                    tempBoard[i][j] = player;
                    root.addNode(tempBoard);
                    tempBoard[i][j] = 0;
                }
            }
        }
    }

    /**
     * Checks to see who, if anyone, won
     *
     * @param board The game board
     * @return Returns 10 if computer wins, -10 if user wins, or 0 is no one wins
     */
    public int check(int[][] board)
    {
        int win, win2, dWin, dWin2;

        dWin = board[0][0] + board[1][1] + board[2][2];
        dWin2 = board[0][2] + board[1][1] + board[2][0];

        for(int i = 0; i < 3; i ++)
        {
            win = win2 = 0;
            for(int j = 0; j < 3; j++)
            {
                win += board[i][j];
                win2 += board[j][i];
            }
            if(win == 3 || win2 == 3 || dWin == 3 || dWin2 == 3)
            {
                return 10; //computer wins
            }
            if(win == 30 || win2 == 30 || dWin == 30 || dWin2 == 30)
            {
                return -10; //user wins
            }
        }
        return 0; //no winner
    }
}


NeuralGame.java

/**
 * Tic Tac Toe game with basic GUI and a working (at least, my idea of one) nerual AI
 * 
 * @author Alex
 * @date 2016
 */
import java.awt.*;
import java.awt.event.*;
import javax.swing.*;
import java.util.*;

public class NeuralGame extends JFrame implements ActionListener
{
    private static final int WIDTH = 500, HEIGHT = 500;
    private JButton[] buttons = new JButton[9];
    private final Dimension gameSize = new Dimension(WIDTH, HEIGHT);
    private int count, choice;
    private int[][] board;
    private boolean user, computer;
    Object[] options = { "Play Again", "Switch who starts", "Exit" };
    Object[] options2 = { "Me", "Computer" };
    Neural network;
    int start;

    /**
     * Creates the window, intializes instance variables, and present you with a dialog for who goes first
     *
     */
    public NeuralGame()
    {
        createWindow();
        network = new Neural();
        board = new int[3][3];
        user = computer = false;
        start = 1;
        int choice = JOptionPane.showOptionDialog(this, "Who is going first?", "Start", 
                JOptionPane.YES_NO_OPTION ,JOptionPane.QUESTION_MESSAGE, null, options2, options[0]);
        if(choice == 1)
        {
            neuralTurn();
            start = 0;
        }
    }

    /**
     * Creates a window with 9 buttons, to represent the tic tac toe gameboard
     *
     */
    public void createWindow() 
    {
        setVisible(true);
        setTitle("Tic Tac Toe");
        setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
        setSize(gameSize);
        setLocationRelativeTo(null);
        getContentPane().setLayout(new GridLayout(3, 3));
        addButtons();

        setPreferredSize(gameSize);
        setMinimumSize(gameSize);
        setMaximumSize(gameSize);

        pack();
    }

    /**
     * Adds buttons to the frame, each of which represents a spot on the tic tac toe board
     *
     */
    public void addButtons()
    {
        for(int i = 0; i < buttons.length; i++)
        {
            buttons[i] = new JButton();
            buttons[i].setFont(new Font(null, Font.PLAIN, 75));
            add(buttons[i]);
            buttons[i].addActionListener(this);
        }
    }

    /**
     * This changes the text of the button for the user, and alters the 2d array to reflect the change
     *
     * @param e A button being presesd
     */
    public void actionPerformed(ActionEvent e)
    {
        JButton button = (JButton) e.getSource();
        choice = Arrays.asList(buttons).indexOf(button);
        if(!button.getText().equals(""))
            return;

        button.setText("O");
        board[choice / 3][choice % 3] = 10;

        if(!user && !computer)
            if(checkWin())
                return;

        if(!full()){
            neuralTurn();
        }

        if(!user && !computer)
            checkWin();
    }

    /**
     * This tests to see if the board is full, in order to prevent errors
     *
     * @return Whether or not the board is full
     */
    public boolean full()
    {
        for(int i = 0; i < board.length; i++)
        {
            for(int j = 0; j < board.length; j++)
            {
                if(board[i][j] == 0)
                {
                    return false;
                }
            }
        }
        return true;
    }

    /**
     * Calls the main TicTacToe class to run the minmax algorithm and make a move for the AI. Interesting note - If the AI can win in multiple ways, it will troll you
     *
     */
    public void neuralTurn()
    {
        setTitle("Thinking...");
        choice = network.getMove(board);
        JButton button = buttons[choice];
        button.setText("X");
        board[choice / 3][choice % 3] = 1;
        setTitle("Tic Tac Toe");
        checkWin();
    }

    /**
     * Checks to see if either player has won
     *
     */
    public void check()
    {
        int win, win2, dWin, dWin2;

        dWin = board[0][0] + board[1][1] + board[2][2];
        dWin2 = board[0][2] + board[1][1] + board[2][0];

        for(int i = 0; i < 3; i ++)
        {
            win = win2 = 0;
            for(int j = 0; j < 3; j++)
            {
                win += board[i][j];
                win2 += board[j][i];
            }
            if(win == 3 || win2 == 3 || dWin == 3 || dWin2 == 3)
            {
                computer = true;
                return;
            }
            if(win == 30 || win2 == 30 || dWin == 30 || dWin2 == 30)
            {
                user = true;
                return;
            }
        }
    }

    /**
     * If someone has won, displays end game popup, asking to either restart, restart with opposing starter, or exit
     *
     */
    public boolean checkWin()
    {
        check();
        String winner1, winner2;
        winner1 = winner2 = "";
        boolean over = false;
        if(user)
        {
            winner1 = "You win!!!";
            winner2 = "You somehow managed to win";
            over = true;
            network.eval(false, 1);
        }
        else if(computer)
        {
            winner1 = "Computer wins";
            winner2 = "The computer won the game";
            over = true;
            network.eval(true, 2);
        }
        else if(full())
        {
            winner1 = "Cat game :(";
            winner2 = "The game was a draw";
            over = true;
            network.eval(true, 1);
        }
        if(over)
        {
            setTitle(winner1);
            int choice = JOptionPane.showOptionDialog(this, winner2, "Result", 
                    JOptionPane.YES_NO_CANCEL_OPTION ,JOptionPane.QUESTION_MESSAGE, null, options, options[0]);
            if(choice == 0)
                resetGame();
            else if(choice == 1)
            {
                resetGame();
                start++;
                start %= 2;
                if(start == 0)
                    neuralTurn();
            }
            else
            {
                this.dispatchEvent(new WindowEvent(this, WindowEvent.WINDOW_CLOSING));
            }
        }
        return over;
    }

    /**
     * Resets the game by setting the board array to 0, and setting all the button texts to empty strings
     *
     */
    public void resetGame()
    {
        user = computer = false;
        setTitle("Tic Tac Toe");
        for(int i = 0; i < buttons.length; i++)
        {
            buttons[i].setText("");
        }
        for(int i = 0; i < board.length; i++)
        {
            for(int j = 0; j < board.length; j++)
                board[i][j] = 0;
        }
    }
}


FileReadWrite.java

/**
 * Write a description of class FileReader here.
 * 
 * @author (your name) 
 * @version (a version number or a date)
 */
import java.io.*;
import java.util.ArrayList;

public class FileReadWrite
{
    BufferedReader br;
    BufferedWriter bw;
    String fileName;
    File file;
    File tempFile;

    public FileReadWrite(String name, boolean append)
    {
        file = new File(name);
        fileName = file.getAbsolutePath();
        file = new File(fileName);
        if(!append)
            file.delete();
    }

    public ArrayList<String> read()
    {
        ArrayList<String> lines = new ArrayList<String>();
        String line;
        try{
            FileReader fr = new FileReader(fileName);
            br = new BufferedReader(fr);
            while((line = br.readLine()) != null)
                lines.add(line);
            fr.close();
        } catch (Exception e) {
            System.out.println("Error");
            e.printStackTrace();
        }

        return lines;
    }

    public String findLine(String word)
    {
        String line;
        try{
            FileReader fr = new FileReader(fileName);
            br = new BufferedReader(fr);
            while((line = br.readLine()) != null)
            {
                String tmp = line.split(" ")[0];
                if(tmp.equals(word))
                {
                    return line;
                }
            }
            fr.close();
        } catch (Exception e) {
            System.out.println("Error");
            e.printStackTrace();
        }

        return "";
    }

    public String findClosest(String word)
    {
        String line;
        String largest = "";
        int longest = 0;
        try{
            FileReader fr = new FileReader(fileName);
            br = new BufferedReader(fr);
            while((line = br.readLine()) != null)
            {
                if(! line.equals(""))
                {
                    int num = getLike(word, line);
                    if(num > longest)
                    {
                        longest = num;
                        largest = line;
                    }
                }
            }
            fr.close();
        } catch (Exception e) {
            System.out.println("Error");
            e.printStackTrace();
        }

        return largest;
    }

    public void write(String ... lines)
    {
        try{
            if (!file.exists())
                file.createNewFile();
            FileWriter fw = new FileWriter(fileName, true);
            bw = new BufferedWriter(fw);
            for(String line : lines)
            {
                bw.write(line);
                bw.newLine();
            }
        } catch (Exception e) {
            System.out.println("Error");
            e.printStackTrace();
        } finally {
            try {
                if (bw != null)
                    bw.close();
            } catch (IOException ex) {
                System.out.println("Error");
                ex.printStackTrace();
            }
        }
    }

    public void replace(String newLine, int choice, int moves, boolean good)
    {
        tempFile = new File(fileName + ".tmp");
        try{
            FileWriter fw = new FileWriter(tempFile);
            bw = new BufferedWriter(fw);
            FileReader fr = new FileReader(fileName);
            br = new BufferedReader(fr);
            String line = "";
            while((line = br.readLine()) != null)
            {
                if(!line.equals(newLine))
                {   
                    bw.write(line);
                    bw.newLine();
                } else {
                    String weights = line.split(" ")[1];
                    String toAdd = line.split(" ")[0] + " ";
                    String nodes[] = weights.split("[)]");
                    int changed = (int) Integer.parseInt(nodes[choice].split("[(]")[1]);
                    if(good)
                    {
                        changed += moves;
                        if(changed > 100)
                            changed = 100;
                    }
                    else
                    {
                        changed -= moves;
                        if(changed < 0)
                            changed = 0;
                    }
                    for(int i = 0; i < nodes.length; i++)
                    {
                        if(i == choice)
                        {
                            toAdd += choice + "(" + changed + ")";
                        } else {
                            toAdd += nodes[i] + ")";
                        }
                    }
                    bw.write(toAdd);
                    bw.newLine();
                }
            }
            fr.close();
        } catch (Exception e) {
            System.out.println("Error");
            e.printStackTrace();
        } finally {
            try {
                if (bw != null)
                    bw.close();
            } catch (IOException ex) {
                System.out.println("Error");
                ex.printStackTrace();
            }
        }
        update();
    }

    public void update()
    {
        System.gc();
        file.setWritable(true);
        file.delete();
        tempFile.renameTo(file);
    }

    public int getLike(String line1, String line2)
    {
        String lines1[] = line1.split("");
        String lines2[] = line2.split("");
        int count = 0;
        for(int i = 0; i < lines1.length; i++)
        {
            if(lines1[i].equals(lines2[i]))
                count++;
        }

        return count;
    }
}


BinaryTreeGame.java

/**
 * Tic Tac Toe game with basic GUI and a working MinMax AI
 * 
 * @author Alex
 * @date 2016
 */
import java.awt.*;
import java.awt.event.*;
import javax.swing.*;
import java.util.*;

public class BinaryTreeGame extends JFrame implements ActionListener
{
    private static final int WIDTH = 500, HEIGHT = 500;
    private JButton[] buttons = new JButton[9];
    private final Dimension gameSize = new Dimension(WIDTH, HEIGHT);
    private int count, choice;
    private int[][] board;
    private boolean user, computer;
    Object[] options = { "Play Again", "Switch who starts", "Exit" };
    Object[] options2 = { "Me", "Computer" };
    TicTacToe tGame;
    int start;
    int runs;

    /**
     * Creates the window, intializes instance variables, and present you with a dialog for who goes first
     *
     */
    public BinaryTreeGame()
    {
        createWindow();
        board = new int[3][3];
        user = computer = false;
        start = 1;
        int choice = JOptionPane.showOptionDialog(this, "Who is going first?", "Start", 
                JOptionPane.YES_NO_OPTION ,JOptionPane.QUESTION_MESSAGE, null, options2, options[0]);
        tGame = new TicTacToe();
        if(choice == 1)
        {
            aiTurn();
            start = 0;
        }
    }

    /**
     * Creates a window with 9 buttons, to represent the tic tac toe gameboard
     *
     */
    public void createWindow() 
    {
        setVisible(true);
        setTitle("Tic Tac Toe");
        setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
        setSize(gameSize);
        setLocationRelativeTo(null);
        getContentPane().setLayout(new GridLayout(3, 3));
        addButtons();

        setPreferredSize(gameSize);
        setMinimumSize(gameSize);
        setMaximumSize(gameSize);

        pack();
    }

    /**
     * Adds buttons to the frame, each of which represents a spot on the tic tac toe board
     *
     */
    public void addButtons()
    {
        for(int i = 0; i < buttons.length; i++)
        {
            buttons[i] = new JButton();
            buttons[i].setFont(new Font(null, Font.PLAIN, 75));
            add(buttons[i]);
            buttons[i].addActionListener(this);
        }
    }

    /**
     * This changes the text of the button for the user, and alters the 2d array to reflect the change
     *
     * @param e A button being presesd
     */
    public void actionPerformed(ActionEvent e)
    {
        JButton button = (JButton) e.getSource();
        choice = Arrays.asList(buttons).indexOf(button);
        if(!button.getText().equals(""))
            return;

        button.setText("O");
        board[choice / 3][choice % 3] = 10;

        if(!user && !computer)
            if(checkWin())
                return;

        if(!full()){
            aiTurn();
        }

        if(!user && !computer)
            checkWin();
    }

    /**
     * This tests to see if the board is full, in order to prevent errors
     *
     * @return Whether or not the board is full
     */
    public boolean full()
    {
        for(int i = 0; i < board.length; i++)
        {
            for(int j = 0; j < board.length; j++)
            {
                if(board[i][j] == 0)
                {
                    return false;
                }
            }
        }
        return true;
    }

    /**
     * Calls the main TicTacToe class to run the minmax algorithm and make a move for the AI. Interesting note - If the AI can win in multiple ways, it will troll you
     *
     */
    public void aiTurn()
    {
        setTitle("Thinking...");
        Node<Integer> node = new Node<Integer>();
        node.data = board;
        tGame.move(node, 0);
        choice = tGame.endGame(node);
        JButton button = buttons[choice];
        button.setText("X");
        board[choice / 3][choice % 3] = 1;
        setTitle("Tic Tac Toe");
        checkWin();
    }

    /**
     * Checks to see if either player has won
     *
     */
    public void check()
    {
        int win, win2, dWin, dWin2;

        dWin = board[0][0] + board[1][1] + board[2][2];
        dWin2 = board[0][2] + board[1][1] + board[2][0];

        for(int i = 0; i < 3; i ++)
        {
            win = win2 = 0;
            for(int j = 0; j < 3; j++)
            {
                win += board[i][j];
                win2 += board[j][i];
            }
            if(win == 3 || win2 == 3 || dWin == 3 || dWin2 == 3)
            {
                computer = true;
                return;
            }
            if(win == 30 || win2 == 30 || dWin == 30 || dWin2 == 30)
            {
                user = true;
                return;
            }
        }
    }

    /**
     * If someone has won, displays end game popup, asking to either restart, restart with opposing starter, or exit
     *
     */
    public boolean checkWin()
    {
        check();
        String winner1, winner2;
        winner1 = winner2 = "";
        boolean over = false;
        if(user)
        {
            winner1 = "You win!!!";
            winner2 = "You somehow managed to win";
            over = true;
        }
        else if(computer)
        {
            winner1 = "Computer wins";
            winner2 = "The computer won the game";
            over = true;
        }
        else if(full())
        {
            winner1 = "Cat game :(";
            winner2 = "The game was a draw";
            over = true;
        }
        if(over)
        {
            setTitle(winner1);
            int choice = JOptionPane.showOptionDialog(this, winner2, "Result", 
                    JOptionPane.YES_NO_CANCEL_OPTION ,JOptionPane.QUESTION_MESSAGE, null, options, options[0]);
            if(choice == 0)
                resetGame();
            else if(choice == 1)
            {
                resetGame();
                start++;
                start %= 2;
                if(start == 0)
                    aiTurn();
            }
            else
            {
                this.dispatchEvent(new WindowEvent(this, WindowEvent.WINDOW_CLOSING));
            }
        }
        return over;
    }

    /**
     * Resets the game by setting the board array to 0, and setting all the button texts to empty strings
     *
     */
    public void resetGame()
    {
        user = computer = false;
        tGame = null;
        tGame = new TicTacToe();
        setTitle("Tic Tac Toe");
        for(int i = 0; i < buttons.length; i++)
        {
            buttons[i].setText("");
        }
        for(int i = 0; i < board.length; i++)
        {
            for(int j = 0; j < board.length; j++)
                board[i][j] = 0;
        }
    }
}
